### PR TITLE
TLS custom certificate validation plugin

### PR DESCRIPTION
**Pull request by Ricfilipe**
As discussed in #4219, currently there is no OCSP/CRL validation for remote peer certificates.
In this PR, a hook is added for a custom certificate validation for remote certificate. The hook is only called when the remote certificate is being verified. The hook will call the plugin with SSL_CONTEXT which allows the custom validation.

A plugin example, called tls-test-validation,  was added that verifies if a certificate has an expected identifier.

```
destination test {
    syslog("172.16.17.4" transport("tls") ip-protocol(4) port(6514) keep-alive(no)
        suppress(5)
        tls(
            ca-dir("/path/to/ca/")
            peer-verify(yes)
        )
        tls-test-validation(identity("test"))
        disk-buffer(
            mem-buf-length(2000)
            disk-buf-size(8388608)
            dir("/tmp")
        )
    );
```

Let me know if any changes are needed

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by Ricfilipe** [comment]
@MrAnno, @bazsi  FYI

**Comment by MrAnno** [comment]
@kira-syslogng ok to test

File: modules/examples/inner-destinations/tls-test-validation/tls-test-validation.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2020 One Identity
+ * Copyright (c) 2020 Laszlo Budai <laszlo.budai@outlook.com>
```
Comment (MrAnno) [inline modules/examples/inner-destinations/tls-test-validation/tls-test-validation.h:3]: You can use your own name in the newly introduces files.
Please also update the year.

**Review by MrAnno — COMMENTED** [review]
This is excellent work. Thank you for working on this PR.

I only had a minor note.

**Comment by bazsi** [comment]
Thanks for creating this.

I like the signal infrastructure, looks good. I was wondering if the "tls-test-validation" plugin could somehow be evolved into a more evolved set of features and moved out of examples. What has come up were stuff like calling an external script to perform validation or even do an OCSP call. Anyway, an example is good first step.

On a related note, I can see it performs a `tls_verify_certificate_name()`, but that is already validated by the afinet driver too, with a different value (host_name). Is there an actual use-case with doing this validation twice, or it's just meant to be used as an example?

**Comment by Ricfilipe** [comment]
The "tls-test-validation" is just meant to be used as an example and demonstrate how to access the X509_STORE_CTX and validate a certificate.

I am planning to create a custom plugin that will communicate with an external certificate validator through a socket based on these [changes ](https://github.com/syslog-ng/syslog-ng/compare/master...amdrsantos:syslog-ng:tls-custom-certificate-validation). 

Since the use case is pretty specific, I was not planning to add the plugin to this PR and instead added the example.

**Comment by MrAnno** [comment]
Thank you.

Your contribution is valid without the OI copyright.
To make it pass to the copyright  check, please add the following:

```diff
diff --git a/tests/copyright/policy b/tests/copyright/policy
index 3da6f2a03..567616b8d 100644
--- a/tests/copyright/policy
+++ b/tests/copyright/policy
@@ -166,6 +166,7 @@ modules/diskq/tests/test_logqueue_disk\.c$
 modules/metrics-probe/metrics-probe(|-parser|-plugin|-grammar)\.(c|h|ym)$
 modules/metrics-probe/label-template\.(c|h)$
 modules/metrics-probe/tests/test_metrics_probe\.c$
+modules/examples/inner-destinations/tls-test-validation
 scl/fortigate/.*\.conf$
 scl/cee/.*\.conf$
 scl/mariadb/.*\.conf$
```

**Review by MrAnno — APPROVED** [review]

**Pull request merged**